Add pg transport and tests

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155810
* __->__ #154653

Add PG transport and tests under `torch/distributed/checkpoint/`

### API:
```python
def send_checkpoint(self, dst_ranks: list[int], state_dict: object) -> None:
def recv_checkpoint(self, src_rank: int) -> object:
```

### Tests:
```
python test/distributed/checkpoint/test_pg_transport.py
```

### Example:
Under `_pg_transport_example.py` (in https://github.com/pytorch/pytorch/pull/155810)
```
torchrun --nproc_per_node=2 -m torch.distributed.checkpoint._pg_transport_example -- --device cuda
```

Differential Revision: [D76044919](https://our.internmc.facebook.com/intern/diff/D76044919)

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

H-Huang committed:
wip

[ghstack-poisoned]

H-Huang committed:
Update on "[wip] dont merge"




cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[wip] dont merge"




cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] Add pg transport"


Add PG transport and tests under `torch/distributed/checkpoint/`

### API:
```python
def send_checkpoint(self, dst_ranks: list[int], state_dict: object) -> None:
def recv_checkpoint(self, src_rank: int) -> object:
```

### Example:
Under `_pg_transport_example.py`
```
torchrun --nproc_per_node=2 -m torch.distributed.checkpoint._pg_transport_example -- --device cuda
```


cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] Add pg transport"


Add PG transport and tests under `torch/distributed/checkpoint/`

### API:
```python
def send_checkpoint(self, dst_ranks: list[int], state_dict: object) -> None:
def recv_checkpoint(self, src_rank: int) -> object:
```

### Example:
Under `_pg_transport_example.py`
```
torchrun --nproc_per_node=2 -m torch.distributed.checkpoint._pg_transport_example -- --device cuda
```

### Tests:
```
python test/distributed/checkpoint/test_pg_transport.py
```

Differential Revision: [D76044919](https://our.internmc.facebook.com/intern/diff/D76044919)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D76044919).

H-Huang committed:
Update on "[WIP] Add pg transport"


Add PG transport and tests under `torch/distributed/checkpoint/`

### API:
```python
def send_checkpoint(self, dst_ranks: list[int], state_dict: object) -> None:
def recv_checkpoint(self, src_rank: int) -> object:
```

### Tests:
```
python test/distributed/checkpoint/test_pg_transport.py
```

### Example:
Under `_pg_transport_example.py` (in https://github.com/pytorch/pytorch/pull/155810)
```
torchrun --nproc_per_node=2 -m torch.distributed.checkpoint._pg_transport_example -- --device cuda
```

Differential Revision: [D76044919](https://our.internmc.facebook.com/intern/diff/D76044919)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

meetv18 reviewed (approved):
Overall LGTM! Just have some small comments/questions. Approving to unblock to land.

H-Huang commented on torch/distributed/checkpoint/_pg_transport.py:
```diff
@@ -0,0 +1,310 @@
+                # if we did a copy we should wait for the work to complete so we
+                # can free the memory to avoid OOMs
+                if original_device == torch.device("cpu"):
+                    for w in work:
+                        w.wait()
+                    work = []
+
+            for w in work:
+                w.wait()
```
thats a good idea, let me try this out and see how it affects the benchmark

H-Huang committed:
Update on "Add pg transport and tests"


Add PG transport and tests under `torch/distributed/checkpoint/`

### API:
```python
def send_checkpoint(self, dst_ranks: list[int], state_dict: object) -> None:
def recv_checkpoint(self, src_rank: int) -> object:
```

### Tests:
```
python test/distributed/checkpoint/test_pg_transport.py
```

### Example:
Under `_pg_transport_example.py` (in https://github.com/pytorch/pytorch/pull/155810)
```
torchrun --nproc_per_node=2 -m torch.distributed.checkpoint._pg_transport_example -- --device cuda
```

Differential Revision: [D76044919](https://our.internmc.facebook.com/intern/diff/D76044919)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i

H-Huang commented:
@pytorchbot merge -i
